app-modules/meeting-center/src/Managers/OutlookCalendarManager.php

```diff
@@ -31,65 +31,281 @@
 
-        // TODO: Implement deleteEvent() method.
+                $request->execute();
+            } else {
+                throw $exception;
+            }
+        }
     }
 
     public function syncEvents(Calendar $calendar, ?Datetime $start = null, ?Datetime $end = null, ?int $perPage = null): void
     {
-        // TODO: Implement syncEvents() method.
+        $providerEvents = collect($this->getEvents($calendar, $start, $end, $perPage));
+
+        $providerEvents
+            ->each(function (Event $providerEvent) use ($calendar) {
+                $userEvent = $calendar->events()->where('provider_id', $providerEvent->getId())->first() ?? $calendar->events()->make();
+
+                $userEvent->fill([
+                    'provider_id' => $providerEvent->getId(),
+                    'title' => $providerEvent->getSubject(),
+                    'description' => $providerEvent->getBodyPreview(),
+                    'starts_at' => Carbon::parse($providerEvent->getStart()->getDateTime(), $providerEvent->getStart()->getTimeZone()),
+                    'ends_at' => Carbon::parse($providerEvent->getEnd()->getDateTime(), $providerEvent->getEnd()->getTimeZone()),
+                    'attendees' => collect($providerEvent->getAttendees())
+                        ->map(fn ($attendee) => $attendee['emailAddress']['address'])
+                        ->prepend($calendar->provider_email),
+                ]);
+
+                if ($userEvent->isDirty()) {
+                    $userEvent->saveQuietly();
+                }
+            });
+
+        $calendar->events()
+            ->whereNull('provider_id')
+            ->each(fn ($event) => $this->createEvent($event));
+
+        $calendar->events()
+            ->whereNotNull('provider_id')
+            ->whereNotIn('provider_id', $providerEvents->map(fn (Event $event) => $event->getId()))
+            ->delete();
     }
 
     public function revokeToken(Calendar $calendar): bool
     {
-        // TODO: Implement revokeToken() method.
+        // There is currently not a way to do this that doesn't require invalidating all refresh_tokens a User has, which would mess with other applications
         return false;
     }
+
+    public function refreshToken(Calendar $calendar): Calendar
+    {
+        $response = Http::asForm()->post(
+            'https://login.microsoftonline.com/' . config('services.azure_calendar.tenant_id') . '/oauth2/token?api-version=v1.0',
+            [
+                'client_id' => config('services.azure_calendar.client_id'),
+                'client_secret' => config('services.azure_calendar.client_secret'),
+                'grant_type' => 'refresh_token',
+                'scope' => ['Calendars.ReadWrite', 'User.Read', 'offline_access'],
+                'refresh_token' => $calendar->oauth_refresh_token,
+            ]
+        );
+
+        if ($response->clientError() || $response->serverError()) {
+            if ($response->status() === 401) {
```

Comment:
What about `Symfony\Component\HttpFoundation\Response::HTTP_UNAUTHORIZED` instead of 401?